# Global Solution - MAI OCEAN




## O nosso grupo optou pelo subdesafio de desenvolver modelos preditivos com machine learning para prever condições ambientais marítimas.

Desenvolver modelos preditivos usando técnicas de Machine Learning para prever condições ambientais
marítimas. O desenvolvimento envolve as etapas de exploração de dados, levantamento de hipótese,
criação e treinamento de modelo com validações e conclusão.

## Pré-processamento de dados

Dataset abrangente sobre conjunto de dados meteorológicos e oceanográficos
- station_id: Identificação da estação de monitoramento (neste caso, "M1").
- longitude: Longitude da estação, medida em graus leste.
- latitude: Latitude da estação, medida em graus norte.
- time: Data e hora da observação, em formato UTC (Tempo Universal Coordenado).
- AtmosphericPressure: Pressão atmosférica medida em milibares (mb).
- WindDirection: Direção do vento medida em graus verdadeiros (de 0 a 360 graus).
- WindSpeed: Velocidade do vento medida em nós (kn).
- Gust: Rajada de vento medida em nós (kn).
- WaveHeight: Altura das ondas medida em metros (m). Valor NaN (Not a Number) indica que não há dados disponíveis.
- WavePeriod: Período das ondas medido em segundos (s). Valor NaN indica que não há dados disponíveis.
- MeanWaveDirection: Direção média das ondas medida em graus verdadeiros. Valor NaN indica que não há dados disponíveis.
- Hmax: Altura máxima das ondas medida em metros (m). Valor NaN indica que não há dados disponíveis.
- AirTemperature: Temperatura do ar medida em graus Celsius (°C).
- DewPoint: Ponto de orvalho medido em graus Celsius (°C). Valor NaN indica que não há dados disponíveis.
- SeaTemperature: Temperatura do mar medida em graus Celsius (°C).
- RelativeHumidity: Umidade relativa medida em porcentagem (%). Valor NaN indica que não há dados disponíveis.
- QC_Flag: Indicador de Controle de Qualidade (Quality Control Flag), onde 1 indica que os dados passaram na verificação de qualidade e 0 indica que não passaram.

Dados meteorológicos e oceanográficos em tempo real coletados da rede de estações Weather Buoy ancoradas na Irlanda. Os parâmetros coletados incluem DateTime (yyyy-mm-ddThh:mm:ss.sss), pressão atmosférica (mbar), temperatura do ar (degreeCelsius), temperatura do ponto de orvalho (degreeCelsius), velocidade do vento (nós), velocidade máxima do vento de rajada (nós) , Direção do Vento (degreeTrue), Temperatura da Superfície do Mar (degreeCelsius), Período das Ondas (segundos), Altura das Ondas (metros) e Umidade Relativa (%). Dados em tempo real disponíveis para M2, M3, M4, M5 e M6. Dados históricos disponíveis para localização espacial M1, FS1 e M4 original. Os usuários do serviço de download podem escolher uma estação, período de tempo, parâmetro(s) e tipo de arquivo de saída. O download avançado permite que um usuário defina uma área de interesse da caixa delimitadora selecionando uma ou mais bóias da rede. 'NaN' ou '-999' descreve dados ausentes ou indisponíveis.

Real time meteorological and oceanographic data collected from the Irish moored Weather Buoy network of stations. Parameters collected include, DateTime (yyyy-mm-ddThh:mm:ss.sss), Atmospheric Pressure (mbar), Air Temperature (degreeCelsius), DewPoint Temperature (degreeCelsius), Wind Speed (knots), Max Gust Wind Speed (knots), Wind Direction (degreeTrue), Sea Surface Temperature (degreeCelsius), Wave Period (seconds), Wave Height (metres) and Relative Humidity (%). Real time data available for M2, M3, M4, M5 and M6. Historical data available for M1, FS1 and original M4 spatial location. Users of the download service can choose a station, time period, parameter(s) and output file type. Advanced download allows a user define a bounding box area of interest selecting one or more buoys from the network. 'NaN' or '-999' describes missing or unavailable data.

In [232]:
import pandas as pd

In [233]:
df = pd.read_csv('https://query.data.world/s/fzhyg3iwyum2omkmdtbhhhgb6b7fyd?dws=00000')

<ipython-input-233-86cd0b6af0f5>:1: DtypeWarning: Columns (1,2,4,5,6,7,8,9,10,11,12,13,14,15) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('https://query.data.world/s/fzhyg3iwyum2omkmdtbhhhgb6b7fyd?dws=00000')


In [234]:
print(df)

       station_id     longitude       latitude                  time  \
0             NaN  degrees_east  degrees_north                   UTC   
1              M1         -11.2        53.1266  2001-02-06T13:00:00Z   
2              M1         -11.2        53.1266  2001-02-06T14:00:00Z   
3              M1         -11.2        53.1266  2001-02-06T15:00:00Z   
4              M1         -11.2        53.1266  2001-02-06T16:00:00Z   
...           ...           ...            ...                   ...   
613388         M6     -15.88135       53.07482  2017-11-28T09:00:00Z   
613389         M2     -5.424046      53.481003  2017-11-28T10:00:00Z   
613390         M4     -9.999136      54.999966  2017-11-28T10:00:00Z   
613391         M5     -6.704336      51.690426  2017-11-28T10:00:00Z   
613392         M6     -15.88135       53.07482  2017-11-28T10:00:00Z   

       AtmosphericPressure WindDirection WindSpeed    Gust WaveHeight  \
0                       mb  degrees_true        kn      kn    

In [235]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 613393 entries, 0 to 613392
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   station_id           613392 non-null  object 
 1   longitude            613393 non-null  object 
 2   latitude             613393 non-null  object 
 3   time                 613393 non-null  object 
 4   AtmosphericPressure  599242 non-null  object 
 5   WindDirection        588582 non-null  object 
 6   WindSpeed            557220 non-null  object 
 7   Gust                 510294 non-null  object 
 8   WaveHeight           538291 non-null  object 
 9   WavePeriod           540546 non-null  object 
 10  MeanWaveDirection    130196 non-null  object 
 11  Hmax                 123785 non-null  object 
 12  AirTemperature       605453 non-null  object 
 13  DewPoint             389331 non-null  object 
 14  SeaTemperature       556618 non-null  object 
 15  RelativeHumidity 

In [236]:
df.head()

station_id     longitude       latitude                  time  \
0        NaN  degrees_east  degrees_north                   UTC   
1         M1         -11.2        53.1266  2001-02-06T13:00:00Z   
2         M1         -11.2        53.1266  2001-02-06T14:00:00Z   
3         M1         -11.2        53.1266  2001-02-06T15:00:00Z   
4         M1         -11.2        53.1266  2001-02-06T16:00:00Z   

  AtmosphericPressure WindDirection WindSpeed  Gust WaveHeight WavePeriod  \
0                  mb  degrees_true        kn    kn          m          s   
1               967.6         270.0     21.98  33.1        NaN        NaN   
2               969.8         270.0     23.93  35.0        NaN        NaN   
3               972.0         270.0     19.07  31.1        NaN        NaN   
4               973.6         270.0     15.95  25.3        NaN        NaN   

  MeanWaveDirection Hmax AirTemperature  DewPoint SeaTemperature  \
0      degrees_true    m       degree_C  degree_C       degree_C   
1               NaN  NaN            8.9       NaN            9.0   
2               NaN  NaN            8.5       NaN            9.0   
3               NaN  NaN            8.9       NaN            9.0   
4               NaN  NaN            9.1       NaN            9.0   

  RelativeHumidity  QC_Flag  
0          percent      NaN  
1              NaN      1.0  
2              NaN      1.0  
3              NaN      1.0  
4              NaN      1.0

In [237]:
 df.drop(df.index[0], inplace=True)

In [238]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 613392 entries, 1 to 613392
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   station_id           613392 non-null  object 
 1   longitude            613392 non-null  object 
 2   latitude             613392 non-null  object 
 3   time                 613392 non-null  object 
 4   AtmosphericPressure  599241 non-null  object 
 5   WindDirection        588581 non-null  object 
 6   WindSpeed            557219 non-null  object 
 7   Gust                 510293 non-null  object 
 8   WaveHeight           538290 non-null  object 
 9   WavePeriod           540545 non-null  object 
 10  MeanWaveDirection    130195 non-null  object 
 11  Hmax                 123784 non-null  object 
 12  AirTemperature       605452 non-null  object 
 13  DewPoint             389330 non-null  object 
 14  SeaTemperature       556617 non-null  object 
 15  RelativeHumidity 

In [239]:
df.head()

station_id longitude latitude                  time AtmosphericPressure  \
1         M1     -11.2  53.1266  2001-02-06T13:00:00Z               967.6   
2         M1     -11.2  53.1266  2001-02-06T14:00:00Z               969.8   
3         M1     -11.2  53.1266  2001-02-06T15:00:00Z               972.0   
4         M1     -11.2  53.1266  2001-02-06T16:00:00Z               973.6   
5         M1     -11.2  53.1266  2001-02-06T18:00:00Z               976.4   

  WindDirection WindSpeed  Gust WaveHeight WavePeriod MeanWaveDirection Hmax  \
1         270.0     21.98  33.1        NaN        NaN               NaN  NaN   
2         270.0     23.93  35.0        NaN        NaN               NaN  NaN   
3         270.0     19.07  31.1        NaN        NaN               NaN  NaN   
4         270.0     15.95  25.3        NaN        NaN               NaN  NaN   
5         270.0     12.06  19.5        NaN        NaN               NaN  NaN   

  AirTemperature DewPoint SeaTemperature RelativeHumidity  QC_Flag  
1            8.9      NaN            9.0              NaN      1.0  
2            8.5      NaN            9.0              NaN      1.0  
3            8.9      NaN            9.0              NaN      1.0  
4            9.1      NaN            9.0              NaN      1.0  
5            8.8      NaN            9.0              NaN      1.0

In [240]:
# df = df.fillna(0)

In [241]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 613392 entries, 1 to 613392
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   station_id           613392 non-null  object 
 1   longitude            613392 non-null  object 
 2   latitude             613392 non-null  object 
 3   time                 613392 non-null  object 
 4   AtmosphericPressure  599241 non-null  object 
 5   WindDirection        588581 non-null  object 
 6   WindSpeed            557219 non-null  object 
 7   Gust                 510293 non-null  object 
 8   WaveHeight           538290 non-null  object 
 9   WavePeriod           540545 non-null  object 
 10  MeanWaveDirection    130195 non-null  object 
 11  Hmax                 123784 non-null  object 
 12  AirTemperature       605452 non-null  object 
 13  DewPoint             389330 non-null  object 
 14  SeaTemperature       556617 non-null  object 
 15  RelativeHumidity 

In [242]:
df.head()

station_id longitude latitude                  time AtmosphericPressure  \
1         M1     -11.2  53.1266  2001-02-06T13:00:00Z               967.6   
2         M1     -11.2  53.1266  2001-02-06T14:00:00Z               969.8   
3         M1     -11.2  53.1266  2001-02-06T15:00:00Z               972.0   
4         M1     -11.2  53.1266  2001-02-06T16:00:00Z               973.6   
5         M1     -11.2  53.1266  2001-02-06T18:00:00Z               976.4   

  WindDirection WindSpeed  Gust WaveHeight WavePeriod MeanWaveDirection Hmax  \
1         270.0     21.98  33.1        NaN        NaN               NaN  NaN   
2         270.0     23.93  35.0        NaN        NaN               NaN  NaN   
3         270.0     19.07  31.1        NaN        NaN               NaN  NaN   
4         270.0     15.95  25.3        NaN        NaN               NaN  NaN   
5         270.0     12.06  19.5        NaN        NaN               NaN  NaN   

  AirTemperature DewPoint SeaTemperature RelativeHumidity  QC_Flag  
1            8.9      NaN            9.0              NaN      1.0  
2            8.5      NaN            9.0              NaN      1.0  
3            8.9      NaN            9.0              NaN      1.0  
4            9.1      NaN            9.0              NaN      1.0  
5            8.8      NaN            9.0              NaN      1.0

In [243]:
for coluna in df.columns:
  if(coluna != "station_id" and coluna != "time"):
    df[coluna] = df[coluna].astype(float)


In [244]:
df.describe()

longitude       latitude  AtmosphericPressure  WindDirection  \
count  613392.000000  613392.000000        599241.000000  588581.000000   
mean       -9.377838      52.879674          1013.552494     206.033596   
std         3.288569       1.295569            12.327422      89.895692   
min       -15.881350      51.215958           909.900000       0.000000   
25%       -10.548261      51.690426          1006.200000     150.000000   
50%        -9.999136      53.074820          1014.800000     220.000000   
75%        -6.704336      53.481003          1022.070000     271.000000   
max        -5.424046      54.999966          1046.400000     360.000000   

           WindSpeed           Gust     WaveHeight     WavePeriod  \
count  557219.000000  510293.000000  538290.000000  540545.000000   
mean       15.204662      21.320918       2.343247       6.140856   
std         6.997032       9.693340       1.553434       1.740916   
min         0.000000       0.000000       0.000000       2.000000   
25%        10.000000      14.089000       1.200000       5.000000   
50%        15.000000      20.000000       2.000000       6.000000   
75%        20.000000      27.000000       3.000000       7.000000   
max        76.506000     200.000000      17.200000      18.000000   

       MeanWaveDirection           Hmax  AirTemperature       DewPoint  \
count      130195.000000  123784.000000   605452.000000  389330.000000   
mean          221.123707       3.118097       11.629430       8.600431   
std            79.128405       2.311975        2.951637       3.932127   
min             0.000000       0.000000        0.264000     -30.000000   
25%           180.000000       1.465000        9.500000       5.800000   
50%           238.000000       2.539000       11.600000       8.900000   
75%           277.000000       4.102000       14.072000      11.600000   
max           359.000000      21.777000       26.172000      47.600000   

       SeaTemperature  RelativeHumidity        QC_Flag  
count   556617.000000     515710.000000  613392.000000  
mean        12.398436         82.373796       0.906880  
std          2.315692          9.760950       0.290601  
min          0.000000          0.000000       0.000000  
25%         10.600000         75.000000       1.000000  
50%         12.200000         83.000000       1.000000  
75%         14.321000         90.000000       1.000000  
max         24.200000        100.000000       1.000000

In [245]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 613392 entries, 1 to 613392
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   station_id           613392 non-null  object 
 1   longitude            613392 non-null  float64
 2   latitude             613392 non-null  float64
 3   time                 613392 non-null  object 
 4   AtmosphericPressure  599241 non-null  float64
 5   WindDirection        588581 non-null  float64
 6   WindSpeed            557219 non-null  float64
 7   Gust                 510293 non-null  float64
 8   WaveHeight           538290 non-null  float64
 9   WavePeriod           540545 non-null  float64
 10  MeanWaveDirection    130195 non-null  float64
 11  Hmax                 123784 non-null  float64
 12  AirTemperature       605452 non-null  float64
 13  DewPoint             389330 non-null  float64
 14  SeaTemperature       556617 non-null  float64
 15  RelativeHumidity 

In [246]:
df.head()

station_id  longitude  latitude                  time  AtmosphericPressure  \
1         M1      -11.2   53.1266  2001-02-06T13:00:00Z                967.6   
2         M1      -11.2   53.1266  2001-02-06T14:00:00Z                969.8   
3         M1      -11.2   53.1266  2001-02-06T15:00:00Z                972.0   
4         M1      -11.2   53.1266  2001-02-06T16:00:00Z                973.6   
5         M1      -11.2   53.1266  2001-02-06T18:00:00Z                976.4   

   WindDirection  WindSpeed  Gust  WaveHeight  WavePeriod  MeanWaveDirection  \
1          270.0      21.98  33.1         NaN         NaN                NaN   
2          270.0      23.93  35.0         NaN         NaN                NaN   
3          270.0      19.07  31.1         NaN         NaN                NaN   
4          270.0      15.95  25.3         NaN         NaN                NaN   
5          270.0      12.06  19.5         NaN         NaN                NaN   

   Hmax  AirTemperature  DewPoint  SeaTemperature  RelativeHumidity  QC_Flag  
1   NaN             8.9       NaN             9.0               NaN      1.0  
2   NaN             8.5       NaN             9.0               NaN      1.0  
3   NaN             8.9       NaN             9.0               NaN      1.0  
4   NaN             9.1       NaN             9.0               NaN      1.0  
5   NaN             8.8       NaN             9.0               NaN      1.0

In [247]:
time = df['time']
df = df.drop(['station_id', 'time'], axis=1)

In [248]:
# media_colunas = dados.mean()

# dados_preenchidos = dados.fillna(media_colunas)


In [249]:
import random

def adicionar_valor_aleatorio(valor, media):
    desvio_padrao = 0.1
    valor_aleatorio = random.uniform(media - desvio_padrao, media + desvio_padrao)
    novo_valor = valor + valor_aleatorio
    return novo_valor

for coluna in df.columns:
  df[coluna] = df[coluna].apply(lambda x: adicionar_valor_aleatorio(x, df[coluna].mean()))


KeyboardInterrupt: 

In [ ]:
df.info()

In [ ]:
df.head()

In [ ]:
df["time"] = time

In [ ]:
df.head()

apagar as colunas que nao vai usar

converter coluna time

FILTRAR DADOS
- O QC_Flag (Sinalizador de Controle de Qualidade) no seu conjunto de dados indica se os dados passaram por uma verificação de qualidade. No seu arquivo, o QC_Flag é 0 ou 1.

Normalmente, em conjuntos de dados desse tipo:

0 geralmente significa que os dados não passaram na verificação de qualidade, indicando possíveis problemas ou imprecisões.
1 geralmente significa que os dados passaram na verificação de qualidade, indicando que são confiáveis.
Com base no seu arquivo, os valores de QC_Flag são todos 1, sugerindo que os dados registrados passaram nas verificações de qualidade e podem ser considerados precisos e confiáveis.

- station_id: Identificação da estação de monitoramento (neste caso, "M1").
- longitude: Longitude da estação, medida em graus leste.
- latitude: Latitude da estação, medida em graus norte.
- time: Data e hora da observação, em formato UTC (Tempo Universal Coordenado).
- AtmosphericPressure: Pressão atmosférica medida em milibares (mb).
- WindDirection: Direção do vento medida em graus verdadeiros (de 0 a 360 graus).
- WindSpeed: Velocidade do vento medida em nós (kn).
- Gust: Rajada de vento medida em nós (kn).
- WaveHeight: Altura das ondas medida em metros (m). Valor NaN (Not a Number) indica que não há dados disponíveis.
- WavePeriod: Período das ondas medido em segundos (s). Valor NaN indica que não há dados disponíveis.
- MeanWaveDirection: Direção média das ondas medida em graus verdadeiros. Valor NaN indica que não há dados disponíveis.
- Hmax: Altura máxima das ondas medida em metros (m). Valor NaN indica que não há dados disponíveis.
- AirTemperature: Temperatura do ar medida em graus Celsius (°C).
- DewPoint: Ponto de orvalho medido em graus Celsius (°C). Valor NaN indica que não há dados disponíveis.
- SeaTemperature: Temperatura do mar medida em graus Celsius (°C).
- RelativeHumidity: Umidade relativa medida em porcentagem (%). Valor NaN indica que não há dados disponíveis.
- QC_Flag: Indicador de Controle de Qualidade (Quality Control Flag), onde 1 indica que os dados passaram na verificação de qualidade e 0 indica que não passaram.
- Observações dos Dados
- Todas as medições da estação M1 ocorreram em 6 de fevereiro de 2001, entre 13:00 e 18:00 UTC.
- A pressão atmosférica aumentou de 967.6 mb para 976.4 mb ao longo do período observado.
- A direção do vento permaneceu constante a 270 graus (indicando um vento vindo de oeste).
- A velocidade do vento variou de 21.98 kn para 12.06 kn.
- Rajadas de vento variaram de 33.1 kn para 19.5 kn.
- As temperaturas do ar variaram entre 8.5°C e 9.1°C.
- A temperatura do mar permaneceu constante em 9.0°C.
- Os valores para WaveHeight, WavePeriod, MeanWaveDirection, Hmax, - - - DewPoint e RelativeHumidity não estão disponíveis (NaN) para todas as observações.
- O QC_Flag é 1 para todas as entradas, indicando que todos os dados passaram nas verificações de qualidade.
Se precisar de mais alguma explicação ou análise sobre este dataset, estou à disposição!



- DateTime (yyyy-mm-ddThh:mm:ss.sss) data e tempo
- Atmospheric Pressure (mbar) pressão atmosferica
- Air Temperature (degreeCelsius) temperatura do ar
- DewPoint Temperature (degreeCelsius) temperatura do ponto de orvalho
- Wind Speed (knots)  velocidade do vento
- Max Gust Wind Speed (knots) velocidade máxima do vento de rajada
- Wind Direction (degreeTrue) Direção do Vento
- Sea Surface Temperature (degreeCelsius) Temperatura da Superfície do Mar
- Wave Period (seconds)  Período das Ondas
- Wave Height (metres) Altura das Ondas
- Relative Humidity (%) Umidade Relativa



---------------------------------------------------------

- station_id -  Identificação da estação de monitoramento (neste caso, "M1").(PODE TIRAR)
- longitude -  Longitude da estação, medida em graus leste.
- latitude - Latitude da estação, medida em graus norte.
- MeanWaveDirection - Direção média das ondas medida em graus verdadeiros.
- Hmax - Altura máxima das ondas medida em metros (m). (pode tirar)
- QC_Flag -  Indicador de Controle de Qualidade (Quality Control Flag), onde 1 indica que os dados passaram na verificação de qualidade e 0 indica que não passaram.

PARA SAIDA: fazer uma média de todos os fatores(features)

## Análise Exploratória